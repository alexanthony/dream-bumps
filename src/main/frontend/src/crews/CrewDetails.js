import React, {PropTypes} from 'react';
import { connect } from 'react-redux';
import {Card, CardTitle, CardText} from 'material-ui/Card';

import { fetchCrew } from './actions';
import { getCrew } from '../rootReducer';

// import dlHorizontal from '../styles/dl.css';
const percentChange = (now, original) => {
  const change = Math.trunc(100 * ( now - original ) / original,2);
  if (change < 0) {
    return `${change}%`;
  }
  return `+${change}%`;
};

const stepChange = (now, original) => {
  // Looks slightly backwards, because lower position is upwards
  if (now - original < 0) {
    return `+${original-now}`;
  }
  return `-${now - original}`;
};

class CrewDetails extends React.Component {
  componentWillMount() {
    if (this.props.crewId) {
      this.props.fetchCrew(this.props.crewId);
    }
  }

  render() {
    const { crew } = this.props;
    if (crew === undefined) {
      return <div>Loading...</div>;
    }
    return (
      <Card>
        <CardTitle title={crew.name} />
        <CardText>
          <div>
            <dl className="dl-horizontal">
              <dt>Start Price</dt>
              <dd>{crew.startPrice}</dd>
            </dl>
            <dl className="dl-horizontal">
              <dt>Current Price</dt>
              <dd>{`${crew.price} (${percentChange(crew.price, crew.startPrice)})`}</dd>
            </dl>
            <dl className="dl-horizontal">
              <dt>Start Position</dt>
              <dd>{crew.startPosition}</dd>
            </dl>
            <dl className="dl-horizontal">
              <dt>Current Position</dt>
              <dd>{`${crew.position} (${stepChange(crew.position, crew.startPosition)})`}</dd>
            </dl>
            <dl className="dl-horizontal">
              <dt>Held</dt>
              <dd>{crew.holding || '_'}</dd>
            </dl>
          </div>
        </CardText>
      </Card>
    );
  }
}

CrewDetails.propTypes = {
  crewId: PropTypes.string,
  fetchCrew: PropTypes.func,
  crew: PropTypes.object
};

const mapStateToProps = (state, ownProps) => ({
  crew: getCrew(state)(ownProps.crewId)
});

export default connect(mapStateToProps, { fetchCrew })(CrewDetails);
