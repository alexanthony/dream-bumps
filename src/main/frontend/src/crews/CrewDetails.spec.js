import React from 'react';
import { shallow } from 'enzyme';
import { CrewDetails } from './CrewDetails';
import toJson from 'enzyme-to-json';

test('Renders properly', () => {
  const fetchCrew = jest.fn();
  const component = shallow(
    <CrewDetails
      fetchCrew={fetchCrew}
      crewId="1"
      crew={{id: 1, price: 100, startPrice: 76.34, position: 5, startPosition: 8}}
    />
  );
  expect(component).toBeDefined();
  expect(toJson(component)).toMatchSnapshot()
  expect(fetchCrew).toHaveBeenCalled();
});
