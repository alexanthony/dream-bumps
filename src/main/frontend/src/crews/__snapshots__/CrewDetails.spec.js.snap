exports[`test Renders properly 1`] = `
<Card
  expandable={false}
  expanded={null}
  initiallyExpanded={false}>
  <CardTitle />
  <CardText>
    <div>
      <dl
        className="dl-horizontal">
        <dt>
          Start Price
        </dt>
        <dd>
          76.34
        </dd>
      </dl>
      <dl
        className="dl-horizontal">
        <dt>
          Current Price
        </dt>
        <dd>
          100 (+30%)
        </dd>
      </dl>
      <dl
        className="dl-horizontal">
        <dt>
          Start Position
        </dt>
        <dd>
          8
        </dd>
      </dl>
      <dl
        className="dl-horizontal">
        <dt>
          Current Position
        </dt>
        <dd>
          5 (+3)
        </dd>
      </dl>
      <dl
        className="dl-horizontal">
        <dt>
          Held
        </dt>
        <dd>
          _
        </dd>
      </dl>
    </div>
  </CardText>
</Card>
`;
